### ISSUE TITLE
Support for external gpg2, corrupted encoding

### ISSUE THREAD

**Issue by silverqx** [git-for-windows/git#1249]
I have tried to setup _gpg.program = gpg2_, but I have problem with encoding and line endings in output in this command _git log --show-signature_.

I attach screenshots of mintty and cmd console output.

I tried to create gpg.conf with _utf8-strings_ and _display-charset utf-8_ [described here](https://www.gnupg.org/documentation/manuals/gnupg/GPG-Configuration-Options.html), but without success.
gpg2 is from gpg4win and I like it, because it remembers passphrase for some time ( with pinentry ).

Default gpg embedded with git works fine.

git for win version: 2.13.3.windows.1 ( x64 )
os: windows 10 x64

![2017-07-28 1](https://user-images.githubusercontent.com/86900/28734719-8fd03698-73e2-11e7-9035-f5d12060cc24.png)
![2017-07-28](https://user-images.githubusercontent.com/86900/28734722-92640772-73e2-11e7-8583-36bfca68345b.png)

**Comment by dscho** [comment]
Next time, please resize your windows so that the screenshots are not *huge* with *mostly black parts*.

Also, please note that gpg2 is not part of Git for Windows, and there is simply too much else to do as to allow me to play with gpg2...

Hopefully other active contributors can help out!

**Comment by szachara** [comment]
Do you know why this happens? Is there needed some chars translations on the git side?

I'm asking this, because, hmm, may be this could be fixed only by some config setting in gpg2.
I wonder where the problem is, thank you for any advices, I don't know where to start investigate.

**Comment by dscho** [comment]
> Do you know why this happens?

I do not even know what is supposed to happen.

**Comment by jkozera** [comment]
I've just submitted a patch for the issue with newlines at https://public-inbox.org/git/20171111160657.11016-1-jerzy.kozera@gmail.com/T/#u (I have not investigated the non-ASCII characters from the second screenshot though.)

**Comment by dscho** [comment]
@jkozera I guess your contribution stalled at some stage, I do not see this commit in Git's `master`.

**Comment by jkozera** [comment]
@dscho you are right, sadly I never managed to apply the code review comments from the git mailing list.

**Comment by dscho** [comment]
> sadly I never managed to apply the code review comments from the git mailing list.

@jkozera Okay, thanks for the clarification.

Back to the original report: In the meantime, Git for Windows ships with GPG v2.x (as a consequence of MSYS2 upgrading to that version). And Git for Windows never officially supported other GPG versions, anyway, so I will close this here ticket.

**Comment by silverqx** [comment]
Bug still exists, can reproduce with Git for Windows with GPG v2 ( which is default version now ) @dscho

**Comment by dscho** [comment]
@silverqx the real question is: how much do you care about the bug?

I ask because as I pointed out: Git for Windows never officially supported other GPG versions than the one it includes.

If you want to change that, I am very open to PRs to make it so.

The experience in this ticket so far, however, suggests to me that you seem to expect me (or at least others) to make that change, and I don't think that that will work. This use case (i.e. configuring an external GPG) is far too fringe to justify my spending a whole lot of time on it.

And it would seem that @jkozera already gave you a head start, all you need to do is to read the reviews, implement the corresponding changes, and submit the patch for review to the Git mailing list (you can open a PR at https://github.com/git/git and use [GitGitGadget](https://gitgitgadget.github.io/) for convenience).

**Comment by silverqx** [comment]
> @silverqx the real question is: how much do you care about the bug?
> 
> I ask because as I pointed out: Git for Windows never officially supported other GPG versions than the one it includes.
> 

It can be reproduced with included gpg, not with external gpg, but with gpg which is included in Git for Windows, I believe it can be reproduced with a default installation.

Sry, I don't have time for this now, maybe in the future.

**Comment by silverqx** [comment]
I updated first comment.

**Comment by silverqx** [comment]
So my mistake, I had still external gpg on path and works good when I removed it, you can let it closed.

**Comment by Cito** [comment]
I can confirm that the problem still exists when using the current Gpg4win 3.1.16, but it does not happen when using the gpg version bundled with the current Git for Windows 2.33.

Also, as far as I understand this is only a matter of bad output; these bad encodings and CR-characters are not *stored* in the repository.

**Comment by silverqx** [comment]
They don't support external gpg on windows, I think this bug should be fixed, fix would be easy for the author of the original code in GFW, it will be a character encoding problem.

**Comment by dscho** [comment]
> I think this bug should be fixed

No you don't. If you truly did, you would fix it.

> fix would be easy for the author of the original code in GFW, it will be a character encoding problem.

That might have been true before this comment: https://github.com/git-for-windows/git/issues/1249#issuecomment-343675396

After this comment was offered, it is just unwillingness to work on addressing the outstanding concerns. Seriously, look at this: https://public-inbox.org/git/20171111160657.11016-1-jerzy.kozera@gmail.com/T/#u

There are _few_ comments, there is _little_ left to do. There is absolutely no need to lean on others to do your bidding.

**Comment by Cito** [comment]
@dscho Maybe you are confusing @silverqx with @jkozera who sent the patch?

**Comment by dscho** [comment]
@Cito no, I am not. Obviously @jkozera did not work on the patch anymore. Which I interpret as them having lost interest in it. But why should @silverqx not take the baton? They are clearly interested in a resolution, so they should have all the motivation necessary to get going. Git is open source for one reason, and for one reason only: so that people who are interested in specific changes can implement them.

Obviously, you don't _have_ to implement those specific changes. If you decide that, that's fine, however, you then don't get to say "I think this bug should be fixed" as if it was anyone _else_'s responsibility to work on this. You then don't have any right to complain. In open source, it is your responsibility to be the change that you want to see. Others have lived up to it, and so can anybody involved in this ticket.

**Comment by jkozera** [comment]
FWIW, my patch was just a rather random result of a one-day Git-focused hackathon, so my interest here is rather limited, especially since I don't use Windows or gpg on daily basis.

**Comment by Cito** [comment]
Ok. One reason why some users do not "take the baton" may be be that they simply expect others to do it, which is not nice. But there can be also other reasons, e.g. someone may not be aquainted enough with the programming language or not have all the other skills that are needed to create a patch, and it would take the person an unreasonable high effort to learn these skills just to create a patch, compared with the existing maintainers and contributers. In this case, of course the user should not *expect* others to fix an issue, but it should be still allowed for them to indicate their interest in a fix.

So in short, to avoid this sort of tension, users of an open source program should not formulate requests as an obligation to developers, and on the other hand developers should not act on the assumptions that all users have the necessary skills and understanding of the internals that they could fix the issues themselves in a reasonable amount of time.

Personally, this issue does not itch me enough to fix it myself, but of course I would not be against anyone fixing it.
